### PR TITLE
JS: Add steps through date-formatting functions

### PR DESCRIPTION
**Pull request by asgerf**
Not sure these will have much impact tbh, but now we've got them covered at least.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/frameworks/DateFunctions.qll [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+/**
+ * Taint step of form: `f -> momentObj.format(f)`
+ *
+ * The format string can use backslash-escaping to include mostly arbitrary text.
+ */
+private class MomentFormatStep extends TaintTracking::AdditionalTaintStep, DataFlow::CallNode {
+  MomentFormatStep() {
+    this = API::moduleImport("moment").getASuccessor*().getMember("format").getACall()
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/DateFunctions.qll:54]: I think `getASuccessor()` is a bit too general.   
If you look at [this LGTM query](https://lgtm.com/query/6188484492070505122/) you can see some stuff that is included by `getASuccessor()`, that I don't think should be included.  

I think a recursive predicate that uses `getAMember()` and `getReturn()` is the way to go.

File: javascript/ql/src/semmle/javascript/frameworks/DateFunctions.qll [hunk 2]
```diff
@@ -0,0 +1,61 @@
+private import javascript
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/DateFunctions.qll:1]: You are missing toplevel qldoc.

File: javascript/ql/src/semmle/javascript/frameworks/DateFunctions.qll [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+private API::Node formatFunction() {
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/DateFunctions.qll:3]: This is `date-fns` specific, but the predicate name doesn't convey that message.  

How about moving the `date-fns` specific classes/predicates into a `DateFns` module?

File: javascript/ql/src/semmle/javascript/frameworks/DateFunctions.qll [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+private API::Node formatFunctionCurried() {
+  result =
+    API::moduleImport(["date-fns/fp", "date-fns/fp/utc"]).getMember(["format", "lightFormat"])
+  or
+  result =
+    API::moduleImport(["date-fns/fp/format", "date-fns/fp/lightFormat", "date-fns/fp/utc/format",
+          "date-fns/fp/utc/lightFormat"])
+}
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/DateFunctions.qll:18]: Looks like you missed the `date-fns/esm` folder. ([Documentation](https://date-fns.org/v2.0.0-alpha.7/docs/ECMAScript-Modules)) 

The `esm` folder also contains a `esm/fp` sub folder.

Comment (asgerf) [reply javascript/ql/src/semmle/javascript/frameworks/DateFunctions.qll]: Good catch, After a closer look it seems the `utc` folder doesn't actually exist in the latest version of date-fns 1.x nor in 2.x, so I removed that at the same time.

**Comment by asgerf** [comment]
I snuck in support for the `dateformat` package as well.

**Comment by asgerf** [comment]
...and `moment-timezone` as well, which seems to re-export the `moment` API with more features.

I better stop here and run and evaluation.

File: change-notes/1.26/analysis-javascript.md [hunk 5]
```diff
@@ -13,6 +13,7 @@
   - [AWS Serverless](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html)
   - [Alibaba Serverless](https://www.alibabacloud.com/help/doc-detail/156876.htm)
   - [debounce](https://www.npmjs.com/package/debounce)
+  - [date-fns](https://www.npmjs.com/package/date-fns)
```
Comment (erik-krogh) [inline change-notes/1.26/analysis-javascript.md:16]:
```suggestion
  - [date-fns](https://www.npmjs.com/package/date-fns)
  - [dateformat](https://www.npmjs.com/package/dateformat)
```

File: change-notes/1.26/analysis-javascript.md [hunk 6]
```diff
@@ -30,6 +31,7 @@
   - [lodash](https://www.npmjs.com/package/lodash)
   - [lodash.debounce](https://www.npmjs.com/package/lodash.debounce)
   - [lodash.throttle](https://www.npmjs.com/package/lodash.throttle)
+  - [moment](https://www.npmjs.com/package/moment)
```
Comment (erik-krogh) [inline change-notes/1.26/analysis-javascript.md:34]:
```suggestion
  - [moment](https://www.npmjs.com/package/moment)
  - [moment-timezone](https://www.npmjs.com/package/moment-timezone)
```

**Review by erik-krogh — COMMENTED** [review]
You're just missing the new libraries from the change-log. 

Otherwise LGTM (assuming the evaluation works out).

**Review by erik-krogh — DISMISSED** [review]

**Comment by asgerf** [comment]
A [smoke-test](https://github.com/github/asgerf-dist-compare-reports/tree/js/date-functions_1604620518096) evaluation looks fine.

Rebased and reverted changes to the 1.26 change notes and added some for 1.27 in the new format ✨

**Review by erik-krogh — APPROVED** [review]

**Pull request merged**